jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/marshalling/ProcessInstanceMarshaller.java

```diff
@@ -140,63 +140,53 @@ public ProcessInstanceMarshaller(ObjectMarshallingStrategy... strategies) {
         }
     }
     
-    public ProcessInstanceDocument marshalProcessInstanceToDocument(ProcessInstance<?> processInstance) {
+    public ProcessInstanceData marshalProcessInstanceToDocument(ProcessInstance<?> processInstance) {
```

Comment:
Perhaps, we should keep method name consistence by using overloaded method. It's a bit of harder to refactor this class. Maybe could try something like this:
`
    public ProcessInstance<?> unmarshallProcessInstance(byte[] data, Process<?> process) {
        Model m = (Model) process.createModel();
        AbstractProcessInstance<?> processInstance = (AbstractProcessInstance<?>) process.createInstance(m);
        return unmarshallProcessInstance(data, process, processInstance);
    }

    public ProcessInstance<?> unmarshallProcessInstance(byte[] data, Process<?> process, AbstractProcessInstance<?> processInstance) {
        return ProcessInstanceMarshaller.unmarshallProcessInstanceWrapperWithStream((d, bais) -> {
            MarshallerReaderContext context = this.createReadContextWithStream.apply(process, bais);
            String processInstanceType = context.stream.readUTF();
            prepareLegacy(processInstance, context, processInstanceType);
            context.close();
            return processInstance;
        }).apply(data);
    }
    public ProcessInstance<?> unmarshallProcessInstance(ProcessInstanceData data, Process<?> process, AbstractProcessInstance<?> processInstance) {
        return ProcessInstanceMarshaller.unmarshallProcessInstanceWrapper(d -> {
            MarshallerReaderContext context = this.createReadContext.apply(process);
            //converting ProcessInstanceDocument to JBPMMessages
            JsonFormat.Parser parser = JsonFormat.parser();
            JBPMMessages.ProcessInstance.Builder builder = JBPMMessages.ProcessInstance.newBuilder();
            parser.merge(data.getlegacyProcessInstance(), builder);
            
            JBPMMessages.ProcessInstance instance = builder.build();
            context.parameterObject = instance;
            for (Map.Entry<String, Integer> entry : data.getStrategies().entrySet()) {
                ObjectMarshallingStrategy strategyObject = context.resolverStrategyFactory.getStrategyObject(entry.getKey());
                if (strategyObject != null) {
                    context.usedStrategies.put(entry.getValue(), strategyObject);
                }
            }

            String processInstanceType = RuleFlowProcess.RULEFLOW_TYPE;
            prepareLegacy(processInstance, context, processInstanceType);
            context.close();
            return processInstance;
        }).apply(data, processInstance);
    }
    
    @FunctionalInterface
    interface ThrowingFunction<T, R, E extends Exception> {
        R apply(T arg) throws E;
    }

    @FunctionalInterface
    interface ThrowingBiFunction<T, U, R, E extends Exception> {
        R apply(T arg1, U arg2) throws E;
    }

    ThrowingFunction<Process<?>, MarshallerReaderContext, IOException> createReadContext = 
        process -> {return new MarshallerReaderContext(Collections.singletonMap(process.id(), ((AbstractProcess<?>) process).legacyProcess()), 
                                                       null, null,null, env, null, true, true);};

    ThrowingBiFunction<Process<?>, ByteArrayInputStream, MarshallerReaderContext, IOException> createReadContextWithStream = 
        (process, bais) -> {return new MarshallerReaderContext(bais,
                                                               Collections.singletonMap(process.id(), ((AbstractProcess<?>)process).legacyProcess()),
                                                               null, null, null, this.env);};
    
    static Function<byte[], ProcessInstance<?>> unmarshallProcessInstanceWrapperWithStream(
        ThrowingBiFunction<byte[], ByteArrayInputStream, ProcessInstance<?>, IOException> fn) {
        return data -> {
            try (ByteArrayInputStream bais = new ByteArrayInputStream((byte[])data)) {
                return fn.apply(data, bais);
            } catch (Exception e) {
                throw new RuntimeException("Error while unmarshalling process instance", e);
            }
        };
    }           

    static BiFunction<ProcessInstanceData, AbstractProcessInstance<?>, ProcessInstance<?>> unmarshallProcessInstanceWrapper(
        ThrowingFunction<ProcessInstanceData, ProcessInstance<?>, IOException> fn) {
        return (data, processInstance) -> {
            try {
                return fn.apply(data);
            }  catch (Exception e) {
                throw new ProcessInstanceUnmarshallingException(processInstance.id(), e, "Error while unmarshalling process instance with id : ");
            }
        };
    }           

    private void prepareLegacy(AbstractProcessInstance<?> processInstance, MarshallerReaderContext context,
            String processInstanceType) throws IOException {
        org.jbpm.marshalling.impl.ProcessInstanceMarshaller marshaller = ProcessMarshallerRegistry.INSTANCE.getMarshaller( processInstanceType );
        org.kie.api.runtime.process.ProcessInstance legacyProcessInstance = marshaller.readProcessInstance(context);
        processInstance.internalSetProcessInstance(legacyProcessInstance);
    }

`

Reply:
@rhtevan Thanks, I am checking on this.